cmake/Modules/FindNCCL.cmake

```diff
@@ -14,32 +14,34 @@
 # install NCCL in the same location as the CUDA toolkit.
 # See https://github.com/caffe2/caffe2/issues/1601
 
-set(NCCL_ROOT_DIR $ENV{NCCL_ROOT_DIR} CACHE PATH "Folder contains NVIDIA NCCL")
+set(NCCL_INCLUDE_DIR $ENV{NCCL_INCLUDE_DIR} CACHE PATH "Folder contains NVIDIA NCCL headers")
+set(NCCL_LIB_DIR $ENV{NCCL_LIB_DIR} CACHE PATH "Folder contains NVIDIA NCCL libraries")
+set(NCCL_VERSION $ENV{NCCL_VERSION} CACHE STRING "Version of NCCL to build with")
+
+list(APPEND NCCL_ROOT ${NCCL_ROOT_DIR} ${CUDA_TOOLKIT_ROOT_DIR})
+# Compatible layer for CMake <3.12. NCCL_ROOT will be accounted in for searching paths and libraries for CMake >=3.12.
+list(APPEND CMAKE_PREFIX_PATH ${NCCL_ROOT})
 
 find_path(NCCL_INCLUDE_DIRS
   NAMES nccl.h
-  HINTS
-  ${NCCL_INCLUDE_DIR}
-  ${NCCL_ROOT_DIR}
-  ${NCCL_ROOT_DIR}/include
-  ${CUDA_TOOLKIT_ROOT_DIR}/include)
+  HINTS ${NCCL_INCLUDE_DIR})
 
-IF ($ENV{USE_STATIC_NCCL})
-  MESSAGE(STATUS "USE_STATIC_NCCL detected. Linking against static NCCL library")
-  SET(NCCL_LIBNAME "libnccl_static.a")
-ELSE()
+if (USE_STATIC_NCCL)
+  MESSAGE(STATUS "USE_STATIC_NCCL is set. Linking with static NCCL library.")
+  SET(NCCL_LIBNAME "nccl_static")
+  if (NCCL_VERSION)  # Prefer the versioned library if a specific NCCL version is specified
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".a.${NCCL_VERSION}" ${CMAKE_FIND_LIBRARY_SUFFIXES})
+  endif()
+else()
   SET(NCCL_LIBNAME "nccl")
-ENDIF()
+  if (NCCL_VERSION)  # Prefer the versioned library if a specific NCCL version is specified
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ".so.${NCCL_VERSION}" ${CMAKE_FIND_LIBRARY_SUFFIXES})
+  endif()
+endif()
 
 find_library(NCCL_LIBRARIES
   NAMES ${NCCL_LIBNAME}
-  HINTS
-  ${NCCL_LIB_DIR}
-  ${NCCL_ROOT_DIR}
-  ${NCCL_ROOT_DIR}/lib
-  ${NCCL_ROOT_DIR}/lib/x86_64-linux-gnu
-  ${NCCL_ROOT_DIR}/lib64
-  ${CUDA_TOOLKIT_ROOT_DIR}/lib64)
+  HINTS ${NCCL_LIB_DIR})
```

Comment:
I'm actually a little confused by what the manual says here; I think in practice this means that we will stop looking in `lib64` location because cmake will only look in that directory if `FIND_LIBRARY_USE_LIB64_PATHS` is set (and I don't think it is set on ordinary platforms). I don't think this matters though, because I checked some of my system nccl installs and they appear to be in non-lib64 directory.

Reply:
I think FIND_LIBRARY_USE_LIB64_PATHS is set when it is proper (such as a multiarch system) to do so. From the document of find_library:

> If the FIND_LIBRARY_USE_LIB64_PATHS global property is set all search paths will be tested as normal, with 64/ appended, and with all matches of lib/ replaced with lib64/. **This property is automatically set for the platforms that are known to need it** if at least one of the languages supported by the project() command is enabled.